File: src/codexbot/logging_utils.py
from __future__ import annotations

import logging
from pathlib import Path
import threading

from .locks import FileLock
from .paths import logs_dir


LOG_MAX_BYTES = 1_000_000
LOG_BACKUP_COUNT = 3

_thread_locks: dict[Path, threading.RLock] = {}
_thread_locks_guard = threading.Lock()


def _thread_lock(path: Path) -> threading.RLock:
    with _thread_locks_guard:
        return _thread_locks.setdefault(path, threading.RLock())


class ProcessSafeRotatingFileHandler(logging.Handler):
    """A bounded file handler safe for hooks and daemon processes sharing a log.

    The log file is opened only while an inter-process lock is held. This keeps
    rotation from racing between short-lived hook processes and the daemon,
    unlike a normal ``RotatingFileHandler`` that keeps an independent stream in
    every process.
    """

    def __init__(
        self,
        filename: str | Path,
        *,
        maxBytes: int = LOG_MAX_BYTES,
        backupCount: int = LOG_BACKUP_COUNT,
        encoding: str = "utf-8",
        lock_timeout: float = 1.0,
    ) -> None:
        super().__init__()
        self.baseFilename = str(Path(filename).resolve())
        self.maxBytes = max(1, int(maxBytes))
        self.backupCount = max(0, int(backupCount))
        self.encoding = encoding
        self.lock_timeout = max(0.0, lock_timeout)
        self._path = Path(self.baseFilename)
        self._lock_path = self._path.with_name(self._path.name + ".lock")

    def _rotate(self) -> None:
        if self.backupCount == 0:
            self._path.unlink(missing_ok=True)
            return

        for index in range(self.backupCount, 0, -1):
            target = self._path.with_name(f"{self._path.name}.{index}")
            source = self._path if index == 1 else self._path.with_name(
                f"{self._path.name}.{index - 1}"
            )
            if target.exists():
                target.unlink()
            if source.exists():
                source.replace(target)

    def _record_bytes(self, record: logging.LogRecord) -> bytes:
        rendered = self.format(record) + "\n"
        data = rendered.encode(self.encoding, errors="replace")
        if len(data) <= self.maxBytes:
            return data

        marker = "... [log record truncated]\n".encode(self.encoding)
        if len(marker) >= self.maxBytes:
            return marker[: self.maxBytes]
        prefix_limit = self.maxBytes - len(marker)
        prefix = data[:prefix_limit].decode(self.encoding, errors="ignore").encode(self.encoding)
        return (prefix + marker)[: self.maxBytes]

    def emit(self, record: logging.LogRecord) -> None:
        try:
            data = self._record_bytes(record)
            thread_lock = _thread_lock(self._path)
            with thread_lock:
                process_lock = FileLock(self._lock_path, timeout=self.lock_timeout)
                if not process_lock.acquire():
                    return
                try:
                    self._path.parent.mkdir(parents=True, exist_ok=True)
                    current_size = self._path.stat().st_size if self._path.exists() else 0
                    if current_size and current_size + len(data) > self.maxBytes:
                        self._rotate()
                    with self._path.open("ab") as handle:
                        handle.write(data)
                        handle.flush()
                finally:
                    process_lock.release()
        except Exception:
            self.handleError(record)


def configure_logging(name: str = "codexbot", *, verbose: bool = False) -> logging.Logger:
    directory = logs_dir()
    directory.mkdir(parents=True, exist_ok=True)

    logger = logging.getLogger(name)
    logger.setLevel(logging.DEBUG if verbose else logging.INFO)
    logger.propagate = False
    if not logger.handlers:
        handler = ProcessSafeRotatingFileHandler(
            directory / "codexbot.log",
            maxBytes=LOG_MAX_BYTES,
            backupCount=LOG_BACKUP_COUNT,
            encoding="utf-8",
        )
        handler.setFormatter(logging.Formatter("%(asctime)s %(levelname)s %(name)s %(message)s"))
        logger.addHandler(handler)
    return logger

File: src/logging_utils.rs
//! Bounded process-safe file logging shared by hook and daemon processes.

use crate::locks::FileLock;
use crate::paths::logs_dir;
use std::collections::HashMap;
use std::fs::{self, OpenOptions};
use std::io::{self, Write};
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex, OnceLock};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

pub const LOG_MAX_BYTES: u64 = 1_000_000;
pub const LOG_BACKUP_COUNT: usize = 3;
const TRUNCATION_MARKER: &[u8] = b"... [log record truncated]\n";

fn thread_locks() -> &'static Mutex<HashMap<PathBuf, Arc<Mutex<()>>>> {
    static LOCKS: OnceLock<Mutex<HashMap<PathBuf, Arc<Mutex<()>>>>> = OnceLock::new();
    LOCKS.get_or_init(|| Mutex::new(HashMap::new()))
}

fn thread_lock(path: &Path) -> Arc<Mutex<()>> {
    let mut locks = thread_locks()
        .lock()
        .unwrap_or_else(|poisoned| poisoned.into_inner());
    locks
        .entry(path.to_path_buf())
        .or_insert_with(|| Arc::new(Mutex::new(())))
        .clone()
}

#[derive(Debug, Clone)]
pub struct ProcessSafeRotatingFileHandler {
    path: PathBuf,
    max_bytes: u64,
    backup_count: usize,
    lock_timeout: Duration,
}

impl ProcessSafeRotatingFileHandler {
    pub fn new(path: impl Into<PathBuf>, max_bytes: u64, backup_count: usize) -> Self {
        let path = path.into();
        let path = if path.is_absolute() {
            path
        } else {
            std::env::current_dir()
                .unwrap_or_else(|_| PathBuf::from("."))
                .join(path)
        };
        Self {
            path,
            max_bytes: max_bytes.max(1),
            backup_count,
            lock_timeout: Duration::from_secs(1),
        }
    }

    pub fn with_lock_timeout(mut self, timeout: Duration) -> Self {
        self.lock_timeout = timeout;
        self
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    fn lock_path(&self) -> PathBuf {
        let name = self
            .path
            .file_name()
            .map(|value| value.to_string_lossy().into_owned())
            .unwrap_or_else(|| "codexbot.log".to_owned());
        self.path.with_file_name(format!("{name}.lock"))
    }

    fn backup_path(&self, index: usize) -> PathBuf {
        let name = self
            .path
            .file_name()
            .map(|value| value.to_string_lossy().into_owned())
            .unwrap_or_else(|| "codexbot.log".to_owned());
        self.path.with_file_name(format!("{name}.{index}"))
    }

    fn rotate(&self) -> io::Result<()> {
        if self.backup_count == 0 {
            match fs::remove_file(&self.path) {
                Ok(()) => return Ok(()),
                Err(error) if error.kind() == io::ErrorKind::NotFound => return Ok(()),
                Err(error) => return Err(error),
            }
        }

        for index in (1..=self.backup_count).rev() {
            let target = self.backup_path(index);
            let source = if index == 1 {
                self.path.clone()
            } else {
                self.backup_path(index - 1)
            };
            match fs::remove_file(&target) {
                Ok(()) => {}
                Err(error) if error.kind() == io::ErrorKind::NotFound => {}
                Err(error) => return Err(error),
            }
            if source.exists() {
                fs::rename(source, target)?;
            }
        }
        Ok(())
    }

    fn record_bytes(&self, record: &str) -> Vec<u8> {
        let mut data = record.as_bytes().to_vec();
        data.push(b'\n');
        let maximum = self.max_bytes as usize;
        if data.len() <= maximum {
            return data;
        }
        if TRUNCATION_MARKER.len() >= maximum {
            return TRUNCATION_MARKER[..maximum].to_vec();
        }

        let prefix_limit = maximum - TRUNCATION_MARKER.len();
        let mut boundary = prefix_limit.min(data.len());
        while boundary > 0 && !record.is_char_boundary(boundary.min(record.len())) {
            boundary -= 1;
        }
        let mut shortened = data[..boundary].to_vec();
        shortened.extend_from_slice(TRUNCATION_MARKER);
        shortened.truncate(maximum);
        shortened
    }

    /// Write one already-formatted record. `Ok(false)` means another process
    /// held the lock beyond the configured timeout and the best-effort record
    /// was intentionally dropped.
    pub fn emit(&self, record: &str) -> io::Result<bool> {
        let data = self.record_bytes(record);
        let local_lock = thread_lock(&self.path);
        let _local_guard = local_lock
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner());
        let mut process_lock = FileLock::new(self.lock_path(), self.lock_timeout);
        if !process_lock.acquire()? {
            return Ok(false);
        }

        let result = (|| {
            if let Some(parent) = self.path.parent() {
                fs::create_dir_all(parent)?;
            }
            let current_size = match fs::metadata(&self.path) {
                Ok(metadata) => metadata.len(),
                Err(error) if error.kind() == io::ErrorKind::NotFound => 0,
                Err(error) => return Err(error),
            };
            if current_size > 0 && current_size.saturating_add(data.len() as u64) > self.max_bytes {
                self.rotate()?;
            }
            let mut handle = OpenOptions::new()
                .create(true)
                .append(true)
                .open(&self.path)?;
            handle.write_all(&data)?;
            handle.flush()
        })();
        let release_result = process_lock.release();
        result.and(release_result).map(|()| true)
    }
}

#[derive(Debug, Clone)]
pub struct ProcessSafeLogger {
    name: String,
    verbose: bool,
    handler: ProcessSafeRotatingFileHandler,
}

impl ProcessSafeLogger {
    fn timestamp() -> String {
        let elapsed = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or(Duration::ZERO);
        format!("{}.{:03}", elapsed.as_secs(), elapsed.subsec_millis())
    }

    pub fn log(&self, level: &str, message: &str) -> io::Result<bool> {
        self.handler.emit(&format!(
            "{} {} {} {}",
            Self::timestamp(),
            level,
            self.name,
            message
        ))
    }

    pub fn debug(&self, message: &str) -> io::Result<bool> {
        if !self.verbose {
            return Ok(false);
        }
        self.log("DEBUG", message)
    }

    pub fn info(&self, message: &str) -> io::Result<bool> {
        self.log("INFO", message)
    }

    pub fn warning(&self, message: &str) -> io::Result<bool> {
        self.log("WARNING", message)
    }

    pub fn error(&self, message: &str) -> io::Result<bool> {
        self.log("ERROR", message)
    }

    pub fn handler(&self) -> &ProcessSafeRotatingFileHandler {
        &self.handler
    }
}

pub fn configure_logging(name: &str, verbose: bool) -> io::Result<ProcessSafeLogger> {
    let directory = logs_dir();
    fs::create_dir_all(&directory)?;
    Ok(ProcessSafeLogger {
        name: name.to_owned(),
        verbose,
        handler: ProcessSafeRotatingFileHandler::new(
            directory.join("codexbot.log"),
            LOG_MAX_BYTES,
            LOG_BACKUP_COUNT,
        ),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn files_and_backups_remain_bounded() {
        let root = std::env::temp_dir().join(format!("codexbot-log-test-{}", std::process::id()));
        let path = root.join("codexbot.log");
        let handler = ProcessSafeRotatingFileHandler::new(&path, 128, 2);
        for _ in 0..30 {
            handler.emit(&"x".repeat(80)).unwrap();
        }
        for candidate in [&path, &handler.backup_path(1), &handler.backup_path(2)] {
            if candidate.exists() {
                assert!(candidate.metadata().unwrap().len() <= 128);
            }
        }
        assert!(!handler.backup_path(3).exists());
        let _ = fs::remove_dir_all(root);
    }
}
